tensorboard/components/tf_backend/urlPathHelpers.ts

```diff
@@ -12,17 +12,42 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
-export function queryEncoder(params?: any): string {
-  // It's important that the keys be sorted, so we always grab the right file
-  // if we are talking to the backend generated by serialze_tensorboard.py
-  if (params == null) {
-    return '';
+/**
+ * A query parameter value can either be a string or a list of strings.
+ * A string `"foo"` is encoded as `key=foo`; a list `["foo", "bar"]` is
+ * encoded as `key=foo&key=bar`.
+ */
+export type QueryValue = string | string[];
+
+/**
+ * Add query parameters to a URL. Values will be URL-encoded. The URL
+ * may or may not already have query parameters. For convenience,
+ * parameters whose value is `undefined` will be dropped.
+ *
+ * For example, the following expressions are equivalent:
+ *
+ *     addParams("http://foo", {a: "1", b: ["2", "3+4"], c: "5"})
+ *     addParams("http://foo?a=1", {b: ["2", "3+4"], c: "5", d: undefined})
+ *     "http://foo?a=1&b=2&b=3%2B4&c=5"
+ */
+export function addParams(
+    baseURL: string,
+    params: {[param: string]: QueryValue}): string {
+  const keys = Object.keys(params).sort().filter(k => params[k] !== undefined);
```

Comment:
Why's this filter needed?

Reply:
For convenience. It lets you write
```ts
addParams(base, {
  run: run,
  limit_attr_size: limit,
  large_attrs_key: limit ? '_key' : undefined,
})
```
which would otherwise have to be written as
```ts
const params = {};
params.run = run;
if (limit !== undefined) {
  params.limit_attr_size = limit.
  params.large_attrs_key = '_key';
}
addParams(base, params);
```
(Note that this filter was already present in `queryEncoder`; I didn't just decide to add it.)

There's a very common related pattern in React. In React, falsy elements are not rendered. So you can write
```jsx
render() {
  const loginEnabled: boolean = this.props.loginEnabled;
  return <div>
    <span>Hello!</span>
    {loginEnabled && <LoginButton />}
  </div>;
}
```
so if `loginEnabled` is `true` then a `LoginButton` will be rendered; otherwise, `false` will be rendered and so there will be no effect.

Reply:
FYI, this was helpful to know.